File: frontend/src/api/analyses.ts
/**
 * Analyses API - Analysis history endpoints
 */
import apiClient from './client'
import {
  AnalysisDetailResponse,
  AnalysisListItemResponse,
  AnalysisStatsResponse,
  GetAnalysesRequest,
} from '../datahelper/analyses.dataHelper'
import { ErrorResponse } from '../datahelper/client.dataHelper'

/**
 * Lấy danh sách analyses với pagination
 * @param limit - Số lượng kết quả (mặc định: 100)
 * @param offset - Vị trí bắt đầu (mặc định: 0)
 * @returns Danh sách analyses
 */
export const getAnalyses = async (
  limit: number = 100,
  offset: number = 0
): Promise<AnalysisListItemResponse[]> => {
  try {
    const params: GetAnalysesRequest = { limit, offset }
    const response = await apiClient.get<AnalysisListItemResponse[]>('/analyses', {
      params,
    })
    return response.data
  } catch (error) {
    throw error as ErrorResponse
  }
}

/**
 * Lấy chi tiết analysis theo ID
 * @param id - ID của analysis
 * @returns Chi tiết analysis
 */
export const getAnalysisById = async (id: number): Promise<AnalysisDetailResponse> => {
  try {
    const response = await apiClient.get<AnalysisDetailResponse>(`/analyses/${id}`)
    return response.data
  } catch (error) {
    throw error as ErrorResponse
  }
}

/**
 * Lấy analysis theo SHA256 hash
 * @param sha256 - SHA256 hash của file
 * @returns Chi tiết analysis
 */
export const getAnalysisBySha256 = async (sha256: string): Promise<AnalysisDetailResponse> => {
  try {
    const response = await apiClient.get<AnalysisDetailResponse>(`/analyses/sha256/${sha256}`)
    return response.data
  } catch (error) {
    throw error as ErrorResponse
  }
}

/**
 * Lấy thống kê analyses
 * @returns Thống kê analyses
 */
export const getAnalysisStats = async (): Promise<AnalysisStatsResponse> => {
  try {
    const response = await apiClient.get<AnalysisStatsResponse>('/analyses/stats/summary')
    return response.data
  } catch (error) {
    throw error as ErrorResponse
  }
}

File: backend/app/api/v1/routes/analyses.py
"""
Analyses endpoints - API quản lý lịch sử phân tích
"""
import os
import sys
from typing import List, Optional
from fastapi import APIRouter, Query, HTTPException

# Thêm project root vào path
project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..'))
if project_root not in sys.path:
    sys.path.insert(0, project_root)

from app.services.analysis_service import AnalysisService
from app.schemas.analysis import AnalysisResponse

router = APIRouter()
analysis_service = AnalysisService()  # Service quản lý analyses

@router.get("")
async def get_analyses(
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0)
):
    """
    Lấy danh sách analyses với pagination
    
    - limit: Số lượng kết quả (1-1000)
    - offset: Vị trí bắt đầu
    
    Returns:
        {
            "items": List[AnalysisResponse],
            "total": int,
            "limit": int,
            "offset": int
        }
    """
    try:
        # Lấy danh sách analyses với pagination
        analyses = await analysis_service.get_all(limit=limit, offset=offset)
        total = await analysis_service.count_all()  # Tổng số analyses
        
        return {
            "items": analyses,
            "total": total,
            "limit": limit,
            "offset": offset
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching analyses: {str(e)}")

@router.get("/{analysis_id}")
async def get_analysis(analysis_id: int):
    """Lấy chi tiết một analysis theo ID"""
    try:
        analysis = await analysis_service.get_by_id(analysis_id)
        if not analysis:
            raise HTTPException(status_code=404, detail="Analysis not found")
        return analysis
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching analysis: {str(e)}")

@router.get("/sha256/{sha256}")
async def get_analysis_by_sha256(sha256: str):
    """Lấy analysis theo hash SHA256"""
    try:
        analysis = await analysis_service.get_by_sha256(sha256)
        if not analysis:
            raise HTTPException(status_code=404, detail="Analysis not found")
        return analysis
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching analysis: {str(e)}")

@router.get("/stats/summary")
async def get_statistics():
    """Lấy thống kê tổng quan: tổng số, malware, file sạch, 24h gần đây"""
    try:
        stats = await analysis_service.get_statistics()
        return stats
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching statistics: {str(e)}")

@router.delete("/{analysis_id}")
async def delete_analysis(analysis_id: int):
    """Xóa một analysis"""
    try:
        # Kiểm tra analysis có tồn tại không
        analysis = await analysis_service.get_by_id(analysis_id)
        if not analysis:
            raise HTTPException(status_code=404, detail="Analysis not found")
        
        # Xóa analysis và các dữ liệu liên quan (yara_matches, ratings)
        deleted = await analysis_service.delete(analysis_id)
        if not deleted:
            raise HTTPException(status_code=500, detail="Failed to delete analysis")
        
        return {"message": "Analysis deleted successfully", "id": analysis_id}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error deleting analysis: {str(e)}")

